native/dlasy2.go

```diff
@@ -0,0 +1,270 @@
+	}
+
+	// 2×2 case: op[TL11 TL12]*[X11 X12] + SGN*[X11 X12]*op[TR11 TR12] = [B11 B12].
+	//             [TL21 TL22] [X21 X22]       [X21 X22]   [TR21 TR22]   [B21 B22]
+	//
+	// Solve equivalent 4×4 system using complete pivoting.
+	// Set pivots less than smin to smin.
+
+	smin := math.Max(math.Abs(tr[0]), math.Abs(tr[1]))
+	smin = math.Max(smin, math.Max(math.Abs(tr[ldtr]), math.Abs(tr[ldtr+1])))
+	smin = math.Max(smin, math.Max(math.Abs(tl[0]), math.Abs(tl[1])))
+	smin = math.Max(smin, math.Max(math.Abs(tl[ldtl]), math.Abs(tl[ldtl+1])))
+	smin = math.Max(eps*smin, smlnum)
+
+	var t16 [16]float64 // t16 is used as a 4×4 column-major matrix.
+	t16[0*4+0] = tl[0] + sgn*tr[0]
+	t16[1*4+1] = tl[ldtl+1] + sgn*tr[0]
+	t16[2*4+2] = tl[0] + sgn*tr[ldtr+1]
+	t16[3*4+3] = tl[ldtl+1] + sgn*tr[ldtr+1]
+	if tranl {
+		t16[0*4+1] = tl[ldtl]
+		t16[1*4+0] = tl[1]
+		t16[2*4+3] = tl[ldtl]
+		t16[3*4+2] = tl[1]
+	} else {
+		t16[0*4+1] = tl[1]
+		t16[1*4+0] = tl[ldtl]
+		t16[2*4+3] = tl[1]
+		t16[3*4+2] = tl[ldtl]
+	}
+	if tranr {
+		t16[0*4+2] = sgn * tr[1]
+		t16[1*4+3] = sgn * tr[1]
+		t16[2*4+0] = sgn * tr[ldtr]
+		t16[3*4+1] = sgn * tr[ldtr]
+	} else {
+		t16[0*4+2] = sgn * tr[ldtr]
+		t16[1*4+3] = sgn * tr[ldtr]
+		t16[2*4+0] = sgn * tr[1]
+		t16[3*4+1] = sgn * tr[1]
+	}
+
+	var btmp [4]float64
+	btmp[0] = b[0]
+	btmp[1] = b[ldb]
+	btmp[2] = b[1]
+	btmp[3] = b[ldb+1]
+
+	// Perform elimination.
+	bi := blas64.Implementation()
+	var jpiv [4]int
+	for i := 0; i < 3; i++ {
+		var (
+			xmax       float64
+			ipsv, jpsv int
+		)
+		for ip := i; ip < 4; ip++ {
+			for jp := i; jp < 4; jp++ {
+				if math.Abs(t16[ip*4+jp]) >= xmax {
+					xmax = math.Abs(t16[ip*4+jp])
+					ipsv = ip
+					jpsv = jp
+				}
+			}
+		}
+		if ipsv != i {
+			bi.Dswap(4, t16[ipsv*4:], 1, t16[i*4:], 1)
+			btmp[ipsv], btmp[i] = btmp[i], btmp[ipsv]
+		}
+		if jpsv != i {
+			bi.Dswap(4, t16[jpsv:], 4, t16[i:], 4)
+		}
+		jpiv[i] = jpsv
+		if math.Abs(t16[i*4+i]) < smin {
+			ok = false
+			t16[i*4+i] = smin
+		}
+		for j := i + 1; j < 4; j++ {
+			t16[j*4+i] /= t16[i*4+i]
+			btmp[j] -= t16[j*4+i] * btmp[i]
+			for k := i + 1; k < 4; k++ {
+				t16[j*4+k] -= t16[j*4+i] * t16[i*4+k]
+			}
+		}
+	}
+	if math.Abs(t16[3*4+3]) < smin {
+		t16[3*4+3] = smin
+	}
+	scale = 1
+	if 8*smlnum*math.Abs(btmp[0]) > math.Abs(t16[0*4+0]) ||
+		8*smlnum*math.Abs(btmp[1]) > math.Abs(t16[1*4+1]) ||
+		8*smlnum*math.Abs(btmp[2]) > math.Abs(t16[2*4+2]) ||
+		8*smlnum*math.Abs(btmp[3]) > math.Abs(t16[3*4+3]) {
+		maxbtmp := math.Max(math.Abs(btmp[0]), math.Abs(btmp[1]))
+		maxbtmp = math.Max(maxbtmp, math.Max(math.Abs(btmp[2]), math.Abs(btmp[3])))
+		scale = 1 / 8 / maxbtmp
+		btmp[0] *= scale
+		btmp[1] *= scale
+		btmp[2] *= scale
+		btmp[3] *= scale
+	}
+	var tmp [4]float64
+	for i := 0; i < 4; i++ {
+		k := 3 - i
+		temp := 1 / t16[k*4+k]
+		tmp[k] = btmp[k] * temp
+		for j := k + 1; j < 4; j++ {
+			tmp[k] -= temp * t16[k*4+j] * tmp[j]
+		}
+	}
+	for i := 0; i < 3; i++ {
+		if jpiv[2-i] != 2-i {
```

Comment:
Maybe a quick note about `2-i` cf `4-1`. (I find these things harder to keep in my head than bit fiddling, which does get a comment).